File: v1/code/go/src/jvmgo/ch02/classpath/entry.go
package classpath

import "os"
import "strings"

// :(linux/unix) or ;(windows)
const pathListSeparator = string(os.PathListSeparator)

type Entry interface {
	// className: fully/qualified/ClassName.class
	readClass(className string) ([]byte, Entry, error)
	String() string
}

func newEntry(path string) Entry {
	if strings.Contains(path, pathListSeparator) {
		return newCompositeEntry(path)
	}

	if strings.HasSuffix(path, "*") {
		return newWildcardEntry(path)
	}

	if strings.HasSuffix(path, ".jar") || strings.HasSuffix(path, ".JAR") ||
		strings.HasSuffix(path, ".zip") || strings.HasSuffix(path, ".ZIP") {

		return newZipEntry(path)
	}

	return newDirEntry(path)
}

File: v1/code/java/jvmgo_java/ch03/src/main/java/com/github/jvmgo/classpath/Entry.java
package com.github.jvmgo.classpath;

import java.io.File;

public interface Entry {

    // className: fully/qualified/ClassName.class
    byte[] readClass(String className) throws Exception;

    // factory method
    static Entry create(String path) {
        if (path.contains(File.pathSeparator)) {
            return new CompositeEntry(path);
        } if (path.endsWith("*")) {
            return new WildcardEntry(path);
        } if (path.endsWith(".jar") || path.endsWith(".JAR") ||
                path.endsWith(".zip") || path.endsWith(".ZIP")) {
            return new ZipEntry(path);
        } else {
            return new DirEntry(path);
        }
    }

}
